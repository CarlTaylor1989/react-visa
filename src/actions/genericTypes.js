export const CHANGE_SCREEN = 'CHANGE_SCREEN';
export const RESET_SCREEN = 'RESET_SCREEN';
export const SUSPEND_DATA_PARSED = 'SUSPEND_DATA_PARSED';
export const SET_GENERIC_DATA = 'SET_GENERIC_DATA';
export const UPDATE_LAUNCHED = 'UPDATE_LAUNCHED';
export const SET_SCREEN_REFERRER = 'SET_SCREEN_REFERRER';
export const SET_POPUP_REFERRER = 'SET_POPUP_REFERRER';
export const SET_PRODUCT_PAGE_INACTIVE = 'SET_PRODUCT_PAGE_INACTIVE';
export const TOGGLE_SETTINGS_BTN_STATE = 'TOGGLE_SETTINGS_BTN_STATE';
export const DISABLE_FEEDBACK = 'DISABLE_FEEDBACK';
export const SET_CLIENT_INTRO_SEEN = 'SET_CLIENT_INTRO_SEEN';
export const SET_NETWORK_INTRO_SEEN = 'SET_NETWORK_INTRO_SEEN';
export const SET_INTRO_SEQ_COMPLETED = 'SET_INTRO_SEQ_COMPLETED';
