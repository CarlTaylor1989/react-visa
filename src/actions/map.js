import * as types from './mapTypes';

export function setChallenges(challenges) {
  return {
    type: types.SET_CHALLENGES,
    challenges
  };
}

export function setFamilies(families) {
  return {
    type: types.SET_FAMILIES,
    families
  };
}

export function initialiseMapState(initialised) {
  return {
    type: types.MAP_STATE_INITIALISED,
    initialised
  };
}

export function updateChallengeStatus(id, familyId, status) {
  return {
    type: types.UPDATE_CHALLENGE_STATUS,
    id,
    familyId,
    status
  };
}

export function updateFamilyStatus(id, status) {
  return {
    type: types.UPDATE_FAMILY_STATUS,
    id,
    status
  };
}

export function setLastChallengesModified(lastChallengesModified) {
  return {
    type: types.SET_LAST_CHALLENGES_MODIFIED,
    lastChallengesModified
  };
}

export function setLastProductCompleted(lastProductCompleted) {
  return {
    type: types.SET_LAST_PRODUCT_COMPLETED,
    lastProductCompleted
  };
}

export function resetLastProductCompleted() {
  return {
    type: types.RESET_LAST_PRODUCT_COMPLETED
  };
}

export function setCurrentFamilyProduct(currentMap, currentFamily, currentProduct) {
  return {
    type: types.SET_CURRENT_FAMILY_PRODUCT,
    currentMap,
    currentFamily,
    currentProduct
  };
}

export function addPerfectChallenge(challenge) {
  return {
    type: types.ADD_PERFECT_CHALLENGE,
    challenge
  };
}

export function setCompletedProducts(products) {
  return {
    type: types.SET_COMPLETED_PRODUCTS,
    products
  };
}

export function addCompletedProduct(productId) {
  return {
    type: types.ADD_COMPLETED_PRODUCT,
    productId
  };
}

export function setCompletedChallenges(challenges) {
  return {
    type: types.SET_COMPLETED_CHALLENGES,
    challenges
  };
}

export function addCompletedChallenge(challengeId) {
  return {
    type: types.ADD_COMPLETED_CHALLENGE,
    challengeId
  };
}

export function resetProgress() {
  return {
    type: types.RESET_PROGRESS
  };
}

export function setProductPageViewedSession() {
  return {
    type: types.SET_PRODUCT_PAGE_VIEWED_SESSION
  };
}
