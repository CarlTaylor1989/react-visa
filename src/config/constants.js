// Loader
export const REMOVE_LOADER_AFTER = 2000;

// Diagnostic
export const DIAGNOSTIC_NOT_STARTED = 0;
export const DIAGNOSTIC_STARTED = 1;
export const DIAGNOSTIC_COMPLETED = 2;

// Client statuses
export const CLIENT_NOT_STARTED = 0;
export const CLIENT_IN_PROGRESS = 1;
export const CLIENT_COMPLETED = 2;
export const CLIENT_LOCKED = 3;
export const CLIENT_GIVENUP = 4;

// Family statuses
export const FAMILY_INCOMPLETE_STATUS = 0;
export const FAMILY_COMPLETION_STATUS = 1;

// Challenge statuses
export const CHALLENGE_UNAVAILABLE_STATUS = -1;
export const CHALLENGE_AVAILABLE_STATUS = 0;
export const CHALLENGE_IN_PROGRESS_STATUS = 1;
export const CHALLENGE_FAILED_STATUS = 2;
export const CHALLENGE_COMPLETION_STATUS = 3;
export const CHALLENGE_PERFECT_STATUS = 4;

// Challenge CSS status classes
export const CHALLENGE_NOT_STARTED_CLASS = 'notStarted';
export const CHALLENGE_AVAILABLE_CLASS = 'available';
export const CHALLENGE_COMPLETED_CLASS = 'completed';
export const CHALLENGE_PERFECT_CLASS = 'perfect';

// Product CSS status classes
export const PRODUCT_INCOMPLETE_CLASS = 'incomplete';
export const PRODUCT_COMPLETED_CLASS = 'completed';

// SCO completion statuses and perfect score
export const SCO_PERFECT_SCORE = 100;
export const SCO_COMPLETION_PASSED = 'passed';
export const SCO_COMPLETION_FAILED = 'failed';
export const SCO_COMPLETION_UNKNOWN = 'unknown';

// Streak bonus
export const ACTIVATE_STREAK_AFTER = 14; // hours
export const MAXIMUM_STREAK = 4;

// Prompt types
export const ACHIEVEMENT_PROMPT = 'achievement';
export const RANK_PROMPT = 'rank';
export const PRODUCT_PROMPT = 'product';
export const BONUS_STREAK_PROMPT = 'bonusStreak';
export const EXIT_SUMMARY_PROMPT_AFTER = 5000;

// Consecutive days
export const CONSECUTIVE_LOWER = 14;
export const CONSECUTIVE_UPPER = 33;

// Tooltip delay
export const DISPLAY_TOOLTIPS_AFTER = 500; // miliseconds

// xAPI
export const XAPI_CHALLENGE_MIN_SCORE = 0;
export const XAPI_CHALLENGE_MAX_SCORE = 2;
export const XAPI_FAIL_CHALLENGE = 'fail';
export const XAPI_FAIL_CHALLENGE_SCORE = 0;
export const XAPI_COMPLETED_CHALLENGE = 'pass';
export const XAPI_COMPLETED_CHALLENGE_SCORE = 1;
export const XAPI_PERFECT_CHALLENGE = 'perfect';
export const XAPI_PERFECT_CHALLENGE_SCORE = 2;
export const XAPI_FEEDBACK_MIN_RATING = 1;
export const XAPI_FEEDBACK_MAX_RATING = 5;

// Cornerstone LMS URL
export const LMS_URL = 'https://visalms.csod.com/LMS/scorm/';

// Feedback
export const FEEDBACK_TRIGGER = 3;
