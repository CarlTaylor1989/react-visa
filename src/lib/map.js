import _ from 'lodash';
import {
  CHALLENGE_UNAVAILABLE_STATUS,
  CHALLENGE_IN_PROGRESS_STATUS,
  CHALLENGE_AVAILABLE_STATUS,
  CHALLENGE_FAILED_STATUS,
  CHALLENGE_COMPLETION_STATUS,
  CHALLENGE_PERFECT_STATUS,
  CHALLENGE_NOT_STARTED_CLASS,
  CHALLENGE_AVAILABLE_CLASS,
  CHALLENGE_COMPLETED_CLASS,
  CHALLENGE_PERFECT_CLASS,
  FAMILY_INCOMPLETE_STATUS,
  FAMILY_COMPLETION_STATUS,
  SCO_COMPLETION_PASSED,
  SCO_COMPLETION_FAILED,
  SCO_COMPLETION_UNKNOWN,
  SCO_PERFECT_SCORE,
  XAPI_FAIL_CHALLENGE,
  XAPI_FAIL_CHALLENGE_SCORE,
  XAPI_COMPLETED_CHALLENGE,
  XAPI_COMPLETED_CHALLENGE_SCORE,
  XAPI_PERFECT_CHALLENGE,
  XAPI_PERFECT_CHALLENGE_SCORE
} from '../config/constants';

export const lib = {};

lib.determineFamilyStatus = (data) => {
  let status;
  if (data.failed || data.inprogress || data.available || data.unavailable) {
    status = FAMILY_INCOMPLETE_STATUS;
  } else {
    status = FAMILY_COMPLETION_STATUS;
  }
  return status;
};

lib.getStoredChallengeStatus = (data) => {
  const challenges = {
    [data.id]: {}
  };

  if (data.perfect) {
    data.perfect.forEach((ch) => {
      challenges[data.id][ch] = CHALLENGE_PERFECT_STATUS;
    });
  }

  if (data.completed) {
    data.completed.forEach((ch) => {
      challenges[data.id][ch] = CHALLENGE_COMPLETION_STATUS;
    });
  }

  if (data.failed) {
    data.failed.forEach((ch) => {
      challenges[data.id][ch] = CHALLENGE_FAILED_STATUS;
    });
  }

  if (data.inprogress) {
    data.inprogress.forEach((ch) => {
      challenges[data.id][ch] = CHALLENGE_IN_PROGRESS_STATUS;
    });
  }

  if (data.available) {
    data.available.forEach((ch) => {
      challenges[data.id][ch] = CHALLENGE_AVAILABLE_STATUS;
    });
  }

  if (data.unavailable) {
    data.unavailable.forEach((ch) => {
      challenges[data.id][ch] = CHALLENGE_UNAVAILABLE_STATUS;
    });
  }

  return challenges;
};

export const getStoredMapData = (suspendData) => {
  const families = [];
  let challenges = {};

  if (suspendData && suspendData.map && suspendData.map.families
    && suspendData.map.families.length) {
    suspendData.map.families.forEach((family) => {
      families.push({
        id: family.id,
        status: lib.determineFamilyStatus(family)
      });

      challenges = {
        ...challenges,
        ...lib.getStoredChallengeStatus(family)
      };
    });
  }

  return {
    families,
    challenges
  };
};

/**
 * Gets the challenge status from the stored data
 * @param {array} challenges A list of stored challenges
 * @param {string} challengeId Challenge ID
 * @param {string} familyId Family ID
 * @param {boolean} isFirst Is the challenge the first in the product line
 * @return {number}
 */
export const getStoredChallengeStatus = (challenges, challengeId, familyId, isFirst) => {
  let status;

  if (challenges[familyId] && challenges[familyId][challengeId] !== undefined) {
    status = challenges[familyId][challengeId];
  } else if (isFirst) {
    status = CHALLENGE_AVAILABLE_STATUS;
  } else {
    status = CHALLENGE_UNAVAILABLE_STATUS;
  }

  return status;
};

/**
 * Determines the challenge status from the SCORM data
 * @param {string} status The completion status
 * @param {number} score The SCO score
 * @return {number}
 */
export const determineChallengeStatus = (status, score) => {
  let challengeStatus = CHALLENGE_AVAILABLE_STATUS;
  if (status === SCO_COMPLETION_PASSED) {
    if (score === SCO_PERFECT_SCORE) {
      challengeStatus = CHALLENGE_PERFECT_STATUS;
    } else {
      challengeStatus = CHALLENGE_COMPLETION_STATUS;
    }
  } else if (status === SCO_COMPLETION_FAILED) {
    challengeStatus = CHALLENGE_FAILED_STATUS;
  } else if (status === SCO_COMPLETION_UNKNOWN) {
    challengeStatus = CHALLENGE_IN_PROGRESS_STATUS;
  }
  return challengeStatus;
};

/**
 * Determines the class of an element and whether it's disabled or not based on the challeng status
 * @param {number} status The current challenge status
 * @return {object} An object with the class name and the disabled status
 */
export const getChallengeStatus = (status) => {
  let className = '';
  let disabled = false;
  switch (status) {
    case CHALLENGE_UNAVAILABLE_STATUS:
      className = CHALLENGE_NOT_STARTED_CLASS;
      disabled = true;
      break;
    case CHALLENGE_AVAILABLE_STATUS:
    case CHALLENGE_IN_PROGRESS_STATUS:
    case CHALLENGE_FAILED_STATUS:
      className = CHALLENGE_AVAILABLE_CLASS;
      break;
    case CHALLENGE_COMPLETION_STATUS:
      className = CHALLENGE_COMPLETED_CLASS;
      break;
    case CHALLENGE_PERFECT_STATUS:
      className = CHALLENGE_PERFECT_CLASS;
      break;
    default:
      className = CHALLENGE_NOT_STARTED_CLASS;
      break;
  }
  return {
    className,
    disabled
  };
};

/**
 * Returns the next available challenge index
 * @param {array} challenges A list of challenges
 * @return {number}
 */
export const getNextChallengeIndex = (challenges) => {
  const found = _.findIndex(challenges, ch => (
    ch.status === CHALLENGE_AVAILABLE_STATUS
    || ch.status === CHALLENGE_IN_PROGRESS_STATUS
    || ch.status === CHALLENGE_FAILED_STATUS
  ));
  return found < 0 ? challenges.length - 1 : found;
};

/**
 * Returns the challenge source data
 * @param {object} challenge The challenge from the game state
 * @param {object} challenges A dictionary of families with list of challenges
 */
export const findChallengeInChallenges = (challenge, challenges) => (
  challenges[challenge.familyId].find(ch => (
    ch.id === challenge.id && ch.productId === challenge.productId
  ))
);

/**
 * Gets the  status and score for the xAPI statement
 * @param {number} challengeStatus The challenge status
 * @return {object}
 */
export const getChallengeXAPIStatus = (status) => {
  const xApiStatus = {
    response: '',
    score: XAPI_FAIL_CHALLENGE_SCORE,
    success: true
  };

  switch (status) {
    case CHALLENGE_COMPLETION_STATUS:
      xApiStatus.response = XAPI_COMPLETED_CHALLENGE;
      xApiStatus.score = XAPI_COMPLETED_CHALLENGE_SCORE;
      break;
    case CHALLENGE_PERFECT_STATUS:
      xApiStatus.response = XAPI_PERFECT_CHALLENGE;
      xApiStatus.score = XAPI_PERFECT_CHALLENGE_SCORE;
      break;
    case CHALLENGE_UNAVAILABLE_STATUS:
    case CHALLENGE_AVAILABLE_STATUS:
    case CHALLENGE_IN_PROGRESS_STATUS:
    case CHALLENGE_FAILED_STATUS:
    default:
      xApiStatus.response = XAPI_FAIL_CHALLENGE;
      xApiStatus.score = XAPI_FAIL_CHALLENGE_SCORE;
      xApiStatus.success = false;
      break;
  }

  return xApiStatus;
};
