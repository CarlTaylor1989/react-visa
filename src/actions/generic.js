import * as types from './genericTypes';

export function changeScreen(screen) {
  return {
    type: types.CHANGE_SCREEN,
    screen
  };
}

export function resetScreen() {
  return {
    type: types.RESET_SCREEN
  };
}

export function suspendDataParsed(dataReady) {
  return {
    type: types.SUSPEND_DATA_PARSED,
    dataReady
  };
}

export function setGenericData(data) {
  return {
    type: types.SET_GENERIC_DATA,
    data
  };
}

export function updateLaunched() {
  return {
    type: types.UPDATE_LAUNCHED
  };
}

export function setScreenReferrer(screenReferrer) {
  return {
    type: types.SET_SCREEN_REFERRER,
    screenReferrer
  };
}

export function setPopupReferrer(popupReferrer) {
  return {
    type: types.SET_POPUP_REFERRER,
    popupReferrer
  };
}

export function setProductPageInactive(productPageInactive) {
  return {
    type: types.SET_PRODUCT_PAGE_INACTIVE,
    productPageInactive
  };
}

export function toggleSettingsBtnState() {
  return {
    type: types.TOGGLE_SETTINGS_BTN_STATE
  };
}

export function disableFeedback() {
  return {
    type: types.DISABLE_FEEDBACK
  };
}

export const setClientIntroSeen = () => ({
  type: types.SET_CLIENT_INTRO_SEEN
});

export const setNetworkIntroSeen = () => ({
  type: types.SET_NETWORK_INTRO_SEEN
});

export const setIntroSeqCompleted = () => ({
  type: types.SET_INTRO_SEQ_COMPLETED
});
