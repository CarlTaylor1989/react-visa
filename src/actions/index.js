import { push } from 'connected-react-router';
import * as AchievementActions from './achievements';
import * as AudioActions from './audio';
import * as ClientActions from './client';
import * as DiagnosticActions from './diagnostic';
import * as GenericActions from './generic';
import * as HintsActions from './hints';
import * as MapActions from './map';
import * as PromptActions from './prompts';
import * as ScoreActions from './score';
import * as ScormActions from './scorm';
import * as SettingsActions from './settings';
import * as StreakActions from './streak';
import * as TutorialActions from './tutorial';

const ActionCreators = Object.assign(
  {},
  AchievementActions,
  AudioActions,
  ClientActions,
  DiagnosticActions,
  GenericActions,
  HintsActions,
  MapActions,
  PromptActions,
  ScoreActions,
  ScormActions,
  SettingsActions,
  StreakActions,
  TutorialActions,
  { push }
);

export default ActionCreators;
