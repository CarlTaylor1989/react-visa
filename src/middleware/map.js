import * as types from '../actions/mapTypes';
import * as mapActions from '../actions/map';
import * as promptActions from '../actions/prompts';
import {
  CHALLENGE_COMPLETION_STATUS,
  CHALLENGE_PERFECT_STATUS
} from '../config/constants';

export const setCompletedChallenges = ({ dispatch, getState }) => next => (action) => {
  if (action.type === types.UPDATE_CHALLENGE_STATUS
    && (action.status === CHALLENGE_COMPLETION_STATUS
      || action.status === CHALLENGE_PERFECT_STATUS)) {
    const mapData = getState().mapData;
    const challengeId = action.familyId + action.id;
    if (!mapData.completedChallenges.includes(challengeId)) {
      dispatch(mapActions.addCompletedChallenge(challengeId));
    }
  }

  if (action.type === types.SET_CHALLENGES) {
    const completedChallenges = action.challenges.filter(ch => (
      ch.status === CHALLENGE_COMPLETION_STATUS || ch.status === CHALLENGE_PERFECT_STATUS));
    const completedChallengeIds = completedChallenges.map(ch => ch.familyId + ch.id);
    dispatch(mapActions.setCompletedChallenges(completedChallengeIds));

    completedChallenges.filter(ch => ch.status === CHALLENGE_PERFECT_STATUS).forEach((ch) => {
      dispatch(mapActions.addPerfectChallenge(ch.familyId + ch.id));
    });
  }

  next(action);
};

export const setCompletedProducts = ({ dispatch, getState }) => next => (action) => {
  const updatedAction = action;

  if (action.type === types.SET_LAST_PRODUCT_COMPLETED) {
    const mapData = getState().mapData;
    const productId = mapData.currentFamily + action.lastProductCompleted;
    if (!mapData.completedProducts.includes(productId)) {
      dispatch(mapActions.addCompletedProduct(productId));
      dispatch(promptActions.addProductPrompt(action.lastProductCompleted));
    } else {
      updatedAction.lastProductCompleted = '';
    }
  }

  if (action.type === types.SET_CHALLENGES) {
    const completedProducts = action.challenges.filter(ch => (
      ch.isFinal
      && (ch.status === CHALLENGE_COMPLETION_STATUS || ch.status === CHALLENGE_PERFECT_STATUS)));
    const completedProductsIds = completedProducts.map(ch => ch.familyId + ch.productId);
    dispatch(mapActions.setCompletedProducts(completedProductsIds));
  }

  next(updatedAction);
};

export const perfectChallenge = ({ dispatch }) => next => (action) => {
  if (action.type === types.UPDATE_CHALLENGE_STATUS && action.status === CHALLENGE_PERFECT_STATUS) {
    dispatch(mapActions.addPerfectChallenge(action.familyId + action.id));
  }

  next(action);
};

export default [
  setCompletedChallenges,
  setCompletedProducts,
  perfectChallenge
];
