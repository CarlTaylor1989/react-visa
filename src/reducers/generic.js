import { createReducer } from 'redux-create-reducer';
import * as types from '../actions/genericTypes';
import {
  CONSECUTIVE_LOWER,
  CONSECUTIVE_UPPER,
  FEEDBACK_TRIGGER
} from '../config/constants';
import launchedWithinDays from '../lib/time';

export const initialState = {
  canDisplayIntroPopup: true,
  clientIntroSeen: false,
  clientIntroSeenSession: false,
  consecutive: 1,
  dataReady: false,
  feedback: {
    days: 1,
    lastUpdatedDate: new Date().getTime(),
    triggered: false,
    disabled: false
  },
  introSeqCompleted: false,
  launched: new Date().getTime(),
  networkIntroSeen: false,
  networkIntroSeenSession: false,
  popupReferrer: '',
  productPageInactive: false,
  screen: 'splash',
  screenReferrer: '',
  settingsBtnEnabled: true
};

export const updateConsecutive = (launched, consecutive) => {
  let lastLaunched = new Date().getTime();
  let streak = consecutive;
  if (launched) {
    const hours = (lastLaunched - launched) / (1000 * 60 * 60);
    if (hours >= CONSECUTIVE_UPPER) {
      streak = initialState.consecutive;
    } else if (hours >= CONSECUTIVE_LOWER && hours <= CONSECUTIVE_UPPER) {
      streak += 1;
    } else {
      lastLaunched = launched;
    }
  }
  return {
    launched: lastLaunched,
    consecutive: streak
  };
};

export const updateFeedback = (feedback) => {
  const SIXTEEN_HOURS = 16 * 60 * 60 * 1000;

  if (
    !feedback.disabled
    && !feedback.triggered
    && new Date().getTime() >= feedback.lastUpdatedDate + SIXTEEN_HOURS
  ) {
    const daysValue = feedback.days + 1;

    return {
      ...feedback,
      days: daysValue,
      triggered: daysValue >= FEEDBACK_TRIGGER,
      lastUpdatedDate: new Date().getTime()
    };
  }

  return {
    ...feedback
  };
};

export const genericData = createReducer(Object.assign(initialState), {
  [types.CHANGE_SCREEN](state, action) {
    return {
      ...state,
      screen: action.screen
    };
  },
  [types.RESET_SCREEN](state) {
    return {
      ...state,
      screen: initialState.screen
    };
  },
  [types.SUSPEND_DATA_PARSED](state, action) {
    return {
      ...state,
      dataReady: action.dataReady
    };
  },
  [types.SET_GENERIC_DATA](state, action) {
    return {
      ...state,
      canDisplayIntroPopup: launchedWithinDays(action.data.launched),
      clientIntroSeen: action.data.clientIntroSeen,
      introSeqCompleted: action.data.introSeqCompleted,
      networkIntroSeen: action.data.networkIntroSeen,
      ...updateConsecutive(action.data.launched, action.data.consecutive),
      feedback: {
        ...updateFeedback({ ...action.data.feedback })
      }
    };
  },
  [types.DISABLE_FEEDBACK](state) {
    return {
      ...state,
      feedback: {
        ...state.feedback,
        disabled: true,
        lastUpdatedDate: new Date().getTime()
      }
    };
  },
  [types.UPDATE_LAUNCHED](state) {
    return {
      ...state,
      ...updateConsecutive(state.launched, state.consecutive)
    };
  },
  [types.SET_SCREEN_REFERRER](state, action) {
    return {
      ...state,
      screenReferrer: action.screenReferrer
    };
  },
  [types.SET_POPUP_REFERRER](state, action) {
    return {
      ...state,
      popupReferrer: action.popupReferrer
    };
  },
  [types.SET_PRODUCT_PAGE_INACTIVE](state, action) {
    return {
      ...state,
      productPageInactive: action.productPageInactive
    };
  },
  [types.TOGGLE_SETTINGS_BTN_STATE](state) {
    return {
      ...state,
      settingsBtnEnabled: !state.settingsBtnEnabled
    };
  },
  [types.SET_CLIENT_INTRO_SEEN](state) {
    return {
      ...state,
      clientIntroSeen: true,
      clientIntroSeenSession: true
    };
  },
  [types.SET_NETWORK_INTRO_SEEN](state) {
    return {
      ...state,
      networkIntroSeen: true,
      networkIntroSeenSession: true
    };
  },
  [types.SET_INTRO_SEQ_COMPLETED](state) {
    return {
      ...state,
      introSeqCompleted: true
    };
  }
});

export default { genericData };
