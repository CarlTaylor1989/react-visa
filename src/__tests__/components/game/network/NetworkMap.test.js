/* global toJson */
import React from 'react'; // eslint-disable-line no-unused-vars
import {
  NetworkMap,
  mapStateToProps,
  mapDispatchToProps
} from '../../../../components/game/network/NetworkMap';
import * as helper from '../../../../lib/map';
import * as actions from '../../../../actions/generic';
import * as types from '../../../../actions/genericTypes';
import { FAMILY_COMPLETION_STATUS } from '../../../../config/constants';
import { sendInteractionData } from '../../../../tracking/xapi/XApiAdapter';

jest.mock('../../../../tracking/xapi/XApiAdapter');

let mockProps;
beforeEach(() => {
  mockProps = {
    canDisplayIntroPopup: false,
    challenges: [],
    checkStreakStatus: jest.fn(),
    commit: jest.fn(),
    initialiseMapState: jest.fn(),
    mapInitialised: false,
    networkIntroSeen: false,
    networkIntroSeenSession: false,
    perfect: [],
    prepareSuspendData: jest.fn(),
    screenReferrer: 'irrelevant',
    setChallenges: jest.fn(),
    setCurrentHintGroup: jest.fn(),
    setFamilies: jest.fn(),
    setNetworkIntroSeen: jest.fn(),
    setPopupReferrer: jest.fn(),
    setScreenReferrer: jest.fn(),
    suspendData: {}
  };
});

test('renders the NetworkMap component', () => {
  const wrapper = shallow(<NetworkMap {...mockProps} />);
  expect(toJson(wrapper, { noKey: false })).toMatchSnapshot();
});

test('method componentDidMount', () => {
  // setup
  const wrapper = shallow(<NetworkMap {...mockProps} />, {
    disableLifecycleMethods: true
  });

  // function under test
  wrapper.instance().componentDidMount();

  // expectations
  expect(mockProps.checkStreakStatus).toHaveBeenCalled();
  expect(mockProps.setCurrentHintGroup).toHaveBeenCalledWith(expect.any(String));
  expect(mockProps.setPopupReferrer).toHaveBeenCalledWith(expect.any(String));
});

describe('method componentDidUpdate()', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<NetworkMap {...mockProps} />);
  });

  test('map with one completed family', () => {
    // setup
    helper.getStoredFamilies = jest.fn().mockReturnValue([{
      id: 'm1f2',
      status: FAMILY_COMPLETION_STATUS
    }]);
    wrapper.setProps({
      mapInitialised: false
    });

    // function under test
    wrapper.update();

    // expectations
    expect(mockProps.setFamilies).toHaveBeenCalled();
    expect(mockProps.setChallenges).toHaveBeenCalled();
    expect(mockProps.initialiseMapState).toHaveBeenCalledWith(true);
  });

  test('map with an invalid completed family', () => {
    // setup
    helper.getStoredFamilies = jest.fn().mockReturnValue([{
      id: 'foo',
      status: FAMILY_COMPLETION_STATUS
    }]);
    wrapper.setProps({
      mapInitialised: false
    });

    // function under test
    wrapper.update();

    // expectations
    expect(mockProps.setFamilies).toHaveBeenCalled();
    expect(mockProps.setChallenges).toHaveBeenCalled();
    expect(mockProps.initialiseMapState).toHaveBeenCalledWith(true);
  });

  test('map without any completed families', () => {
    // setup
    helper.getStoredFamilies = jest.fn().mockReturnValue([]);
    wrapper.setProps({
      mapInitialised: false
    });

    // function under test
    wrapper.update();

    // expectations
    expect(mockProps.setFamilies).toHaveBeenCalled();
    expect(mockProps.setChallenges).toHaveBeenCalled();
    expect(mockProps.initialiseMapState).toHaveBeenCalledWith(true);
  });

  test('map already initialised', () => {
    // setup
    wrapper.setProps({
      mapInitialised: true
    });

    // function under test
    wrapper.update();

    // expectations
    expect(mockProps.setFamilies).not.toHaveBeenCalled();
    expect(mockProps.setChallenges).not.toHaveBeenCalled();
    expect(mockProps.initialiseMapState).not.toHaveBeenCalled();
  });
});

test('method componentWillUnmount', () => {
  // setup
  const wrapper = shallow(<NetworkMap {...mockProps} />);
  wrapper.instance().sendPageStatement = jest.fn();

  // function under test
  wrapper.instance().componentWillUnmount();

  // expectations
  expect(wrapper.instance().sendPageStatement).toHaveBeenCalled();
  expect(mockProps.setScreenReferrer).toHaveBeenCalledWith(expect.any(String));
});

test('method sendPageStatement()', () => {
  // setup
  const wrapper = shallow(<NetworkMap {...mockProps} />, {
    disableLifecycleMethods: true
  });
  wrapper.instance().timeAvailable = new Date();

  // function under test
  wrapper.instance().sendPageStatement();

  // expectations
  expect(sendInteractionData).toHaveBeenCalledWith(expect.objectContaining({
    type: 'experienced',
    activity: expect.objectContaining({
      id: 'map',
      type: 'module',
      title: expect.any(String),
      description: expect.any(String)
    }),
    response: expect.objectContaining({
      timeAvailable: expect.any(Date),
      timeResponse: expect.any(Date)
    }),
    parent: '',
    referrer: expect.any(String)
  }));
});

test('should mapStateToProps', () => {
  const state = {
    genericData: {
      screenReferrer: 'irrelevant',
    },
    mapData: {
      initialised: false
    },
    scormData: {
      suspendData: {}
    }
  };
  const props = mapStateToProps(state);
  expect(props).toEqual({
    mapInitialised: state.mapData.initialised,
    screenReferrer: state.genericData.screenReferrer,
    suspendData: state.scormData.suspendData
  });
});

test('should mapDispatchToProps', () => {
  const screen = 'construction';
  const dispatch = jest.fn();
  jest.spyOn(actions, 'changeScreen').mockReturnValue('changeScreen()');
  const props = mapDispatchToProps(dispatch);
  props.changeScreen(screen);
  expect(dispatch).toHaveBeenCalledWith({
    type: types.CHANGE_SCREEN,
    screen
  });
});

describe('method displayPopup()', () => {
  test('networkIntroSeenSession is false', () => {
    // setup
    const wrapper = shallow(<NetworkMap {...mockProps} />, {
      disableLifecycleMethods: true
    });

    // function under test
    const output = wrapper.instance().displayPopup();

    // expectation
    expect(output).toEqual(true);
  });

  test('networkIntroSeenSession is true', () => {
    // setup
    mockProps.networkIntroSeenSession = true;
    const wrapper = shallow(<NetworkMap {...mockProps} />, {
      disableLifecycleMethods: true
    });

    // function under test
    const output = wrapper.instance().displayPopup();

    // expectation
    expect(output).toEqual(false);
  });

  test('networkIntroSeen is true', () => {
    // setup
    mockProps.networkIntroSeen = true;
    const wrapper = shallow(<NetworkMap {...mockProps} />, {
      disableLifecycleMethods: true
    });

    // function under test
    const output = wrapper.instance().displayPopup();

    // expectation
    expect(output).toEqual(true);
  });
});

test('method onClosePopup()', () => {
  // setup
  const wrapper = shallow(<NetworkMap {...mockProps} />, {
    disableLifecycleMethods: true
  });

  // function under test
  wrapper.instance().onClosePopup();

  // expectation
  expect(mockProps.setNetworkIntroSeen).toHaveBeenCalled();
});
