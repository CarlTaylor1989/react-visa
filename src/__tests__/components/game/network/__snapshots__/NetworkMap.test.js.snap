// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the NetworkMap component 1`] = `
<div
  className="networkMap"
>
  <IntroSequencePopup
    onClose={[Function]}
    open={true}
    translations="network"
  />
  <div
    className="contours"
  />
  <div
    className="mapContainer"
  >
    <Connect(FamilyContainer)
      familyId="m1f1"
      key="m1f10"
      mapId="m1"
    />
    <Connect(FamilyContainer)
      familyId="m1f2"
      key="m1f21"
      mapId="m1"
    />
    <Connect(FamilyContainer)
      familyId="m1f3"
      key="m1f32"
      mapId="m1"
    />
    <Connect(FamilyContainer)
      familyId="m1f4"
      key="m1f43"
      mapId="m1"
    />
    <Connect(FamilyContainer)
      familyId="m1f5"
      key="m1f54"
      mapId="m1"
    />
    <span
      className="home"
    />
  </div>
  <PerfectChallenges
    current={0}
    total={0}
  />
  <Minimap />
</div>
`;
